llvm/lib/CodeGen/ReplaceWithVeclib.cpp

```diff
@@ -69,88 +69,97 @@ Function *getTLIFunction(Module *M, FunctionType *VectorFTy,
-  CalltoReplace.getOperandBundlesAsDefs(OpBundles);
-  CallInst *Replacement = IRBuilder.CreateCall(TLIVecFunc, Args, OpBundles);
-  CalltoReplace.replaceAllUsesWith(Replacement);
+  if (CI)
+    CI->getOperandBundlesAsDefs(OpBundles);
+
+  auto *Replacement = IRBuilder.CreateCall(TLIVecFunc, Args, OpBundles);
+  I.replaceAllUsesWith(Replacement);
   // Preserve fast math flags for FP math.
   if (isa<FPMathOperator>(Replacement))
-    Replacement->copyFastMathFlags(&CalltoReplace);
+    Replacement->copyFastMathFlags(&I);
 }
 
-/// Returns true when successfully replaced \p CallToReplace with a suitable
-/// function taking vector arguments, based on available mappings in the \p TLI.
-/// Currently only works when \p CallToReplace is a call to vectorized
-/// intrinsic.
+/// Returns true when successfully replaced \p I with a suitable function taking
+/// vector arguments, based on available mappings in the \p TLI. Currently only
+/// works when \p I is a call to vectorized intrinsic or the frem instruction.
 static bool replaceWithCallToVeclib(const TargetLibraryInfo &TLI,
-                                    CallInst &CallToReplace) {
-  if (!CallToReplace.getCalledFunction())
-    return false;
-
-  auto IntrinsicID = CallToReplace.getCalledFunction()->getIntrinsicID();
-  // Replacement is only performed for intrinsic functions.
-  if (IntrinsicID == Intrinsic::not_intrinsic)
-    return false;
-
-  // Compute arguments types of the corresponding scalar call. Additionally
-  // checks if in the vector call, all vector operands have the same EC.
-  ElementCount VF = ElementCount::getFixed(0);
-  SmallVector<Type *> ScalarArgTypes;
-  for (auto Arg : enumerate(CallToReplace.args())) {
-    auto *ArgTy = Arg.value()->getType();
-    if (isVectorIntrinsicWithScalarOpAtArg(IntrinsicID, Arg.index())) {
-      ScalarArgTypes.push_back(ArgTy);
-    } else if (auto *VectorArgTy = dyn_cast<VectorType>(ArgTy)) {
-      ScalarArgTypes.push_back(ArgTy->getScalarType());
-      // Disallow vector arguments with different VFs. When processing the first
-      // vector argument, store it's VF, and for the rest ensure that they match
-      // it.
-      if (VF.isZero())
-        VF = VectorArgTy->getElementCount();
-      else if (VF != VectorArgTy->getElementCount())
+                                    Instruction &I) {
+  std::string ScalarName;
+  ElementCount EC = ElementCount::getFixed(0);
+  Function *FuncToReplace = nullptr;
+  SmallVector<Type *, 8> ScalarArgTypes;
+  // Compute the argument types of the corresponding scalar call, the scalar
+  // function name, and EC. For calls, it additionally checks if in the vector
+  // call, all vector operands have the same EC.
+  if (auto *CI = dyn_cast<CallInst>(&I)) {
+    Intrinsic::ID IID = CI->getCalledFunction()->getIntrinsicID();
+    assert(IID != Intrinsic::not_intrinsic && "Not an intrinsic");
+    FuncToReplace = CI->getCalledFunction();
+    for (auto Arg : enumerate(CI->args())) {
```

Comment:
it just occurred to me that this loop checks only input arguments, but in case of intrinsic we do not check the return type EC, maybe it is worth to initialise EC to it instead of doing:

"ElementCount EC = ElementCount::getFixed(0);"?
I believe we won't support intrinsics returning scalar value or void at the moment

Reply:
True, it'll have to be a vector, unless VFABI at some point encodes return types.

Such a change would require further refactoring:
1. `isSupportedInstruction` should no longer accept any instructions that do not return a vector (for both calls, frem)
2. setting EC for frem and the assertion can be hoisted here from the else statement (disccused in [prev comment](https://github.com/llvm/llvm-project/pull/76166#discussion_r1442185949))
3. for calls, this would mean that the EC has to be the return type. So, we'll now only check that any vector arguments matches that.

Reply:
@paschalis-mpeis  I think it need to be slightly smarter for CallInst, as AFAIK if the scalar function returns void we expect the vector to also return void, otherwise it has to be a vector.